torch/utils/_benchmark/utils/valgrind_wrapper/timer_interface.py

```diff
@@ -0,0 +1,448 @@
+            f_stdout_stderr = open(stdout_stderr_log, "wb")
+            try:
+                invocation = subprocess.run(
+                    args,
+                    stdout=f_stdout_stderr,
+                    stderr=subprocess.STDOUT,
+                    **kwargs,
+                )
+                with open(stdout_stderr_log, "rt") as f:
+                    return invocation, f.read()
+            finally:
+                f_stdout_stderr.close()
+
+        try:
+            with open(script_file, "wt") as f:
+                f.write(self._construct_script(
+                    stmt=stmt, setup=setup, number=number,
+                    num_threads=num_threads, error_log=error_log,
+                    stat_log=stat_log))
+
+            valgrind_invocation, valgrind_invocation_output = run([
+                "valgrind",
+                "--tool=callgrind",
+                f"--callgrind-out-file={callgrind_out}",
+                "--dump-line=yes",
+                "--dump-instr=yes",
+                "--collect-jumps=yes",
+                "--instr-atstart=yes",
+                "--collect-atstart=no",
+                "python",
+                script_file,
+            ])
+
+            if valgrind_invocation.returncode:
+                error_report = ""
+                if os.path.exists(error_log):
+                    with open(error_log, "rt") as f:
+                        error_report = f.read()
+                if not error_report:
+                    error_report = "Unknown error.\n" + valgrind_invocation_output
+
+                raise OSError(f"Failed to collect callgrind profile:\n{error_report}")
+
+            def parse_output(inclusive: bool):
+                annotate_invocation, annotate_invocation_output = run([
+                    "callgrind_annotate",
+                    f"--inclusive={'yes' if inclusive else 'no'}",
+                    callgrind_out
+                ], check=True)
+
+                begin_collecting = False
+                fn_counts = []
+                for l in annotate_invocation_output.splitlines(keepends=False):
+                    if not begin_collecting and re.match(r"Ir\s+file:function", l):
+                        begin_collecting = True
+                        continue
+
+                    count_match = re.match(r"^\s*([0-9,]+)\s+(.+:.+)$", l)
+                    if count_match:
+                        ir_str, file_function = count_match.groups()
+                        ir = int(ir_str.replace(",", ""))
+                        fn_counts.append((ir, file_function))
+                        continue
+
+                    if begin_collecting and re.match(r"-+", l):
+                        continue
+
+                    begin_collecting = False
+
+                return fn_counts
+            return parse_output(inclusive=True), parse_output(inclusive=False)
+        finally:
+            shutil.rmtree(working_dir)
```

Comment:
In my experience, for debugging purposes, it is almost always preferable to have some way of keeping the working dir around, so you can, e.g., see what the actual script that was being run was.

Reply:
I generally don't like leaving an unbounded number of files around. Particularly since it includes a callgrind output, so it wouldn't take too many failures to potentially run someone out of disk. I'll see if it's straightforward to keep the last `n` around.

Reply:
Therefore the "have some way", and not by default ;)

Reply:
I just came across `torch._appdirs.user_cache_dir`, which looks to be exactly what I wanted. If all runs share the same scratch dir, then I can make reasonable decisions about what to keep when.